piptools/resolver.py

```diff
@@ -25,27 +30,31 @@ class RequirementSummary:
     Summary of a requirement's properties for comparison purposes.
     """
 
-    def __init__(self, ireq):
+    def __init__(self, ireq: InstallRequirement) -> None:
         self.req = ireq.req
         self.key = key_from_ireq(ireq)
         self.extras = frozenset(ireq.extras)
         self.specifier = ireq.specifier
 
-    def __eq__(self, other):
-        return (
-            self.key == other.key
-            and self.specifier == other.specifier
-            and self.extras == other.extras
-        )
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, self.__class__):
+            return (
+                self.key == other.key
+                and self.specifier == other.specifier
+                and self.extras == other.extras
+            )
+        return NotImplemented
```

Comment:
FYI: this line left uncovered, so I've added it to `exclude_lines` to the coverage report config in a619d0dd2f34897b37ed48efff8b5a6aaa967d7e

Reply:
Why not add tests instead?

Reply:
Okay, let's add one.

Reply:
Added in cabb979.